modules/caas/backend/src/main/java/io/cattle/platform/loop/EndpointUpdateLoop.java

```diff
@@ -42,15 +50,19 @@ public Result run(List<Object> input) {
         Map<Long, String> agentIps = new HashMap<>();
         Map<Long, ServiceInfo> services = metadata.getServices().stream().collect(toMap(ServiceInfo::getId, (x) -> x));
         Map<Long, Set<PortInstance>> servicePorts = new HashMap<>();
-        Map<Long, Set<PortInstance>> hostPorts = new HashMap<>();
-
-        for (HostInfo hostInfo : metadataManager.getMetadataForCluster(metadata.getClusterId()).getHosts()) {
+        Metadata clusterMetadata = metadataManager.getMetadataForCluster(metadata.getClusterId());
+        for (HostInfo hostInfo : clusterMetadata.getHosts()) {
             if (StringUtils.isNotBlank(hostInfo.getAgentIp())) {
                 agentIps.put(hostInfo.getId(), hostInfo.getAgentIp());
             }
         }
 
+        boolean updated = false;
+        List<String> validStates = Arrays.asList(InstanceConstants.STATE_RUNNING, InstanceConstants.STATE_STARTING, InstanceConstants.STATE_RESTARTING);
```

Comment:
Just save these as static vars (and as a Set<String>, not List).  It's just wasted cycles to build the list everytime.